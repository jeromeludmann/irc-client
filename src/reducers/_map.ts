import { Action } from "redux";

export const mapReducer = <S, E = void>(map: {
  [action: string]: (state: S, action: Action, extraStates?: E) => S;
}) => (state: S, action: Action, extraStates?: E): S =>
  map.hasOwnProperty(action.type)
    ? map[action.type](state, action, extraStates)
    : state;
