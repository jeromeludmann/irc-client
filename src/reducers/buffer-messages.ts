import {
  CONNECTION_FAILED,
  ConnectionFailedAction,
  RAW_MESSAGES_RECEIVED,
  RawMessagesReceivedAction,
} from "@app/actions/network";
import {
  JoinAction,
  ServerNoticeAction,
  ServerPingAction,
  ChannelPrivmsgAction,
  UserPrivmsgAction,
  JOIN,
  SERVER_PING,
  CHANNEL_PRIVMSG,
  USER_PRIVMSG,
  SERVER_NOTICE,
} from "@app/actions/message-in";

export type MessageListState = string[];

export type MessageListAction =
  | JoinAction
  | ServerNoticeAction
  | ServerPingAction
  | ChannelPrivmsgAction
  | UserPrivmsgAction
  | ConnectionFailedAction
  | RawMessagesReceivedAction;

export const messagesInitialState: MessageListState = [];

export default (
  messages = messagesInitialState,
  action: MessageListAction,
): MessageListState => {
  switch (action.type) {
    case JOIN: {
      const { user, channel } = action.payload;
      const msg = `${user.nick} has joined ${channel}`;
      return [...messages, msg];
    }
    case SERVER_PING: {
      return [...messages, "Ping? (Pong!)"];
    }
    case CHANNEL_PRIVMSG:
    case USER_PRIVMSG: {
      const { user, text } = action.payload;
      const msg = `${user.nick}: ${text}`;
      return [...messages, msg];
    }
    case SERVER_NOTICE: {
      return [...messages, action.payload.text];
    }
    case CONNECTION_FAILED: {
      return [...messages, action.payload.message];
    }
    case RAW_MESSAGES_RECEIVED: {
      return [...messages, ...action.payload.messages];
    }
    default:
      return messages;
  }
};
