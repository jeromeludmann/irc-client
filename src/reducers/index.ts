import {
  RouteState,
  routeInitialState,
  reduceRoute,
} from "@app/reducers/route";
import {
  ServersState,
  serversInitialState,
  reduceServers,
} from "@app/reducers/servers";
import { RoutedAction } from "@app/Route";
import { Reducer } from "redux";

export interface AppState {
  readonly servers: ServersState;
  readonly route: RouteState;
}

export const rootInitialState = {
  servers: serversInitialState,
  route: routeInitialState,
};

export const reduce: Reducer<AppState, RoutedAction> = (
  state = rootInitialState,
  action,
) => ({
  servers: reduceServers(state.servers, action, { route: state.route }),
  route: reduceRoute(state.route, action, { servers: state.servers }),
});
