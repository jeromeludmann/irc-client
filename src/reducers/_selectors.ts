import { createSelector } from "reselect";
import { RootState } from "@app/reducers";
import { ServersState } from "@app/reducers/servers";
import { RouteState } from "@app/reducers/route";

export const selectServers = ({ servers }: RootState): ServersState => servers;

export const selectRoute = ({ route }: RootState): RouteState => route;

export const selectServer = createSelector(
  selectServers,
  selectRoute,
  (servers, { serverKey }) => servers[serverKey],
);
