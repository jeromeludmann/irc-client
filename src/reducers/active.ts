import { Route, STATUS_WINDOW } from "@app/Route";
import { SwitchWindowAction, SWITCH_WINDOW } from "@app/actions/ui-window";
import { JoinAction, JOIN } from "@app/actions/message-in";
import { UserState } from "@app/reducers/user";

export type ActiveWindowState = Route;

export type ActiveWindowAction = JoinAction | SwitchWindowAction;

export const activeInitialState: ActiveWindowState = {
  server: "serverKey", // TODO multi servers
  window: STATUS_WINDOW,
};

type ExtraParams = {
  user: UserState;
};

export default (
  active = activeInitialState,
  action: ActiveWindowAction,
  { user }: ExtraParams,
): ActiveWindowState => {
  switch (action.type) {
    case JOIN:
      return action.payload.user.nick === user.nick ? action.route : active;

    case SWITCH_WINDOW:
      return action.route;

    default:
      return active;
  }
};
