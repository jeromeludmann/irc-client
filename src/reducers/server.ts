import { Action } from "redux";
import reduceModes, {
  ModesState,
  modesInitialState,
} from "@app/reducers/server-modes";
import { WindowState } from "@app/reducers/window";
import reduceBufferRouter, {
  BufferRouterState,
  bufferRouterInitialState,
  BufferRouterAction,
} from "@app/reducers/server-buffers";
import { Route } from "@app/Route";
import { UserState } from "@app/reducers/user";

export type ServerState = {
  readonly modes: ModesState;
  readonly buffers: BufferRouterState;
};

export type ServerAction = Action;

interface ExtraParams {
  user: UserState;
  route: Route;
  active: WindowState;
}

export const serverInitialState: ServerState = {
  modes: modesInitialState,
  buffers: bufferRouterInitialState,
};

export default (
  server = serverInitialState,
  action: ServerAction,
  params: ExtraParams,
): ServerState => ({
  modes: reduceModes(server.modes),
  buffers: reduceBufferRouter(
    server.buffers,
    action as BufferRouterAction,
    params,
  ),
});
