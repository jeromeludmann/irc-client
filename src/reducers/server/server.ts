import reduceModes, {
  ModesState,
  modesInitialState,
} from "@app/reducers/server/modes";
import { WindowState } from "@app/reducers/window";
import reduceBufferRouter, {
  BufferRouterState,
  bufferRouterInitialState,
  BufferRouterAction,
} from "@app/reducers/server/buffer-router";
import { Route } from "@app/Route";
import { UserState } from "@app/reducers/user";
import { RPL_MYINFO, IncomingReplyMyInfoAction } from "@app/actions/incoming";
import { Action } from "redux";

export type ServerState = {
  readonly name: string; // TODO
  readonly availableModes: { user: string[]; channel: string[] }; // TODO
  readonly modes: ModesState;
  readonly buffers: BufferRouterState;
};

export type ServerAction = Action;

interface ExtraParams {
  user: UserState;
  route: Route;
  active: WindowState;
}

export const serverInitialState: ServerState = {
  name: "<unknown>", // TODO
  availableModes: { user: [], channel: [] }, // TODO
  modes: modesInitialState,
  buffers: bufferRouterInitialState,
};

// TODO extract me
function reduceName(name: string, action: IncomingReplyMyInfoAction) {
  switch (action.type) {
    case RPL_MYINFO:
      return action.payload.serverName;
    default:
      return name;
  }
}

// TODO extract me
function reduceAvailableModes(
  availableModes: any,
  action: IncomingReplyMyInfoAction,
) {
  switch (action.type) {
    case RPL_MYINFO:
      return {
        user: action.payload.availableUserModes,
        channel: action.payload.availableChannelModes,
      };
    default:
      return availableModes;
  }
}

export default (
  server = serverInitialState,
  action: ServerAction,
  params: ExtraParams,
): ServerState => ({
  name: reduceName(server.name, action as IncomingReplyMyInfoAction),
  availableModes: reduceAvailableModes(
    server.availableModes,
    action as IncomingReplyMyInfoAction,
  ),
  modes: reduceModes(server.modes),
  buffers: reduceBufferRouter(
    server.buffers,
    action as BufferRouterAction,
    params,
  ),
});
