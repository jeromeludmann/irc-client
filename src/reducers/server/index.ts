import {
  ServerModesState,
  serverModesInitialState,
  reduceServerModes,
} from "@app/reducers/server/modes";
import { RouteState } from "@app/reducers/route";
import {
  ChannelsState,
  channelsInitialState,
  reduceChannels,
} from "@app/reducers/server/channels";
import { RoutedAction } from "@app/Route";
import {
  serverNameInitialState,
  reduceServerName,
} from "@app/reducers/server/name";
import {
  reduceAvailableServerModes,
  AvailableServerModesState,
  availableServerModesInitialState,
} from "@app/reducers/server/availableModes";
import { ServerNameState } from "@app/reducers/server/name";
import {
  UserState,
  userInitialState,
  reduceUser,
} from "@app/reducers/server/user";

export type ServerState = {
  readonly name: ServerNameState;
  readonly user: UserState;
  readonly availableModes: AvailableServerModesState;
  readonly modes: ServerModesState;
  readonly channels: ChannelsState;
};

export const serverInitialState: ServerState = {
  name: serverNameInitialState,
  user: userInitialState,
  availableModes: availableServerModesInitialState,
  modes: serverModesInitialState,
  channels: channelsInitialState,
};

export const reduceServer = (
  server = serverInitialState,
  action: RoutedAction,
  extraStates: { route: RouteState },
): ServerState => ({
  name: reduceServerName(server.name, action),
  user: reduceUser(server.user, action),
  availableModes: reduceAvailableServerModes(server.availableModes, action),
  modes: reduceServerModes(server.modes),
  channels: reduceChannels(server.channels, action, {
    ...extraStates,
    user: server.user,
  }),
});
