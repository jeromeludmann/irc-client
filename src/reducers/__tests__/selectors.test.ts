import {
  selectServers,
  selectRoute,
  selectServer,
  selectServerName,
  selectUser,
  selectServerLag,
  selectUserModes,
  selectAvailableModes,
  selectBuffer,
  selectBuffers,
  selectActivity,
  selectInput,
  selectInputValue,
  selectInputDirtyValue,
  selectInputHistory,
  selectMessages,
} from '@app/reducers/selectors'
import { RootState } from '@app/reducers'
import { serverInitialState } from '@app/reducers/server'
import { BufferKey } from '@app/utils/Route'

describe('selectors', () => {
  const initialState: RootState = {
    servers: {
      serverKey: serverInitialState,
    },
    route: { serverKey: 'serverKey', bufferKey: BufferKey.STATUS },
  }

  it('should select servers', () => {
    expect(selectServers(initialState)).toMatchSnapshot()
  })

  it('should select route', () => {
    expect(selectRoute(initialState)).toMatchSnapshot()
  })

  it('should select server', () => {
    expect(selectServer(initialState)).toMatchSnapshot()
  })

  it('should select server name', () => {
    expect(selectServerName(initialState)).toMatchSnapshot()
  })

  it('should select user', () => {
    expect(selectUser(initialState)).toMatchSnapshot()
  })

  it('should select server lag', () => {
    expect(selectServerLag(initialState)).toMatchSnapshot()
  })

  it('should select user modes', () => {
    expect(selectUserModes(initialState)).toMatchSnapshot()
  })

  it('should select available modes', () => {
    expect(selectAvailableModes(initialState)).toMatchSnapshot()
  })

  it('should select buffers', () => {
    expect(selectBuffers(initialState)).toMatchSnapshot()
  })

  it('should select buffer', () => {
    expect(selectBuffer(initialState)).toMatchSnapshot()
  })

  it('should select activity', () => {
    expect(selectActivity(initialState)).toMatchSnapshot()
  })

  it('should select input', () => {
    expect(selectInput(initialState)).toMatchSnapshot()
  })

  it('should select input value', () => {
    expect(selectInputValue(initialState)).toMatchSnapshot()
  })

  it('should select input dirty value', () => {
    expect(selectInputDirtyValue(initialState)).toMatchSnapshot()
  })

  it('should select input history', () => {
    expect(selectInputHistory(initialState)).toMatchSnapshot()
  })

  it('should select input history', () => {
    expect(selectMessages(initialState)).toMatchSnapshot()
  })
})
