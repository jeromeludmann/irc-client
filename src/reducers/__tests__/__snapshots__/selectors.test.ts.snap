// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors should select activity 1`] = `false`;

exports[`selectors should select available modes 1`] = `
Object {
  "channel": Array [],
  "user": Array [],
}
`;

exports[`selectors should select buffer 1`] = `
Object {
  "activity": false,
  "input": Object {
    "dirtyValue": "",
    "history": Object {
      "index": 0,
      "values": Array [],
    },
    "value": "",
  },
  "messages": Array [],
}
`;

exports[`selectors should select buffers 1`] = `
Object {
  "@raw": Object {
    "activity": false,
    "input": Object {
      "dirtyValue": "",
      "history": Object {
        "index": 0,
        "values": Array [],
      },
      "value": "",
    },
    "messages": Array [],
  },
  "@status": Object {
    "activity": false,
    "input": Object {
      "dirtyValue": "",
      "history": Object {
        "index": 0,
        "values": Array [],
      },
      "value": "",
    },
    "messages": Array [],
  },
}
`;

exports[`selectors should select input 1`] = `
Object {
  "dirtyValue": "",
  "history": Object {
    "index": 0,
    "values": Array [],
  },
  "value": "",
}
`;

exports[`selectors should select input dirty value 1`] = `""`;

exports[`selectors should select input history 1`] = `
Object {
  "index": 0,
  "values": Array [],
}
`;

exports[`selectors should select input history 2`] = `Array []`;

exports[`selectors should select input value 1`] = `""`;

exports[`selectors should select route 1`] = `
Object {
  "bufferKey": "@status",
  "serverKey": "serverKey",
}
`;

exports[`selectors should select server 1`] = `
Object {
  "buffers": Object {
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
  },
  "lag": 0,
  "modes": Object {
    "available": Object {
      "channel": Array [],
      "user": Array [],
    },
    "user": Array [],
  },
  "name": "<unknown>",
  "user": Object {
    "nick": "default_nick",
    "real": "default_name",
    "user": "default_user",
  },
}
`;

exports[`selectors should select server lag 1`] = `0`;

exports[`selectors should select server name 1`] = `"<unknown>"`;

exports[`selectors should select servers 1`] = `
Object {
  "serverKey": Object {
    "buffers": Object {
      "@raw": Object {
        "activity": false,
        "input": Object {
          "dirtyValue": "",
          "history": Object {
            "index": 0,
            "values": Array [],
          },
          "value": "",
        },
        "messages": Array [],
      },
      "@status": Object {
        "activity": false,
        "input": Object {
          "dirtyValue": "",
          "history": Object {
            "index": 0,
            "values": Array [],
          },
          "value": "",
        },
        "messages": Array [],
      },
    },
    "lag": 0,
    "modes": Object {
      "available": Object {
        "channel": Array [],
        "user": Array [],
      },
      "user": Array [],
    },
    "name": "<unknown>",
    "user": Object {
      "nick": "default_nick",
      "real": "default_name",
      "user": "default_user",
    },
  },
}
`;

exports[`selectors should select user 1`] = `
Object {
  "nick": "default_nick",
  "real": "default_name",
  "user": "default_user",
}
`;

exports[`selectors should select user modes 1`] = `Array []`;
